Add a warding processor

micaelbergeron opened the pull request:
~~Please consider this as a draft~~

This processor may be used to track down warding events, such as
Counterwarding, Placement or Expiry. It uses the combat log events
correlated with entity life state tracking.

Sample log output:

```
-- Ward placed
{"time":13,"type":"obs","key":"[78, 80]","slot":0,"x":78,"y":80,"z":130,"entityleft":false,"ehandle":4195051}
{"time":11,"type":"sen","key":"[78, 80]","slot":0,"x":78,"y":80,"z":130,"entityleft":false,"ehandle":4850406}

-- Ward expired
{"time":251,"type":"sen_left","key":"[78, 80]","slot":0,"x":78,"y":80,"z":130,"entityleft":true,"ehandle":4850406}
{"time":373,"type":"obs_left","key":"[78, 80]","slot":0,"x":78,"y":80,"z":130,"entityleft":true,"ehandle":4195051}

-- Ward countered
{"time":22,"type":"obs_left","key":"[78, 82]","slot":0,"attackername":"npc_dota_hero_abaddon","x":78,"y":82,"z":130,"entityleft":true,"ehandle":14631244}
{"time":50,"type":"sen_left","key":"[78, 80]","slot":0,"attackername":"npc_dota_hero_abaddon","x":78,"y":80,"z":130,"entityleft":true,"ehandle":9208152}
```

I still need to test for controlled minions counterwarding.

Feel free to add your thoughts.

Commit:
Add a warding processor

this processor may be used to track down warding events, such as
Counterwarding, Placement or Expiry. It uses the combat log events
correlated with entity life state tracking.

Commit:
Add a warding processor

this processor may be used to track down warding events, such as
Counterwarding, Placement or Expiry. It uses the combat log events
correlated with entity life state tracking.

spheenik reviewed (approved):
I like a lot that finally there's someone using the SpawnsAndDeaths example to good use! ;)

I might be porting the SpawnsAndDeaths processor from clarity-examples to clarity, so that it's events would be available to everyone. Then your Wards-Processor could be simplified. 

One thing that came to my mind: What happens if a ward dies of old age, and another one gets countered on the same tick, and the update of the lifestate for the one dying of old age comes first?

howardchung commented on pom.xml:
```diff
@@ -71,24 +71,52 @@
                     </execution>
                 </executions>
             </plugin>
+            <!-- for some reason the shade-plugin doesn't work well with clarity processors.
```
preferably we should find a way to avoid needing to swap this dependency.

micaelbergeron replied:
I took like 3 hours today trying to make this work. Maybe @spheenik can give some guidance because the shaded jar will crash upon populating the processors. It seems it cannot find the annotations or whatever.

howardchung replied:
IIRC this produces a jar with a different name.  We have to make sure the output file is named the same, or update the Dockerfile to start the correct file.

spheenik replied:
The problem here is that there is an annotation scanner in clarity (`org.atteo.classindex`) that scans annotations on compile time, and produces a file `META-INF/annotations/skadistats.clarity.event.Provides` which contains a list of all classes with Provides-annotations. If you use shade, you got to make sure this file is present and correctly merged.

howardchung commented on src/main/java/yasp/Main.java:
```diff
@@ -2,35 +2,61 @@
+    public static void main(String[] args) {
+	try { 
+                       
+	    if (args.length > 0 && args[0].equals("--file")) { System.exit(parse(args[1])); }
```
Fix brace spacing

micaelbergeron replied:
To be honest, if you want to include the Main class, I'd rather have a proper arg-parsing, like Apache Commons CLI or similar.

howardchung replied:
The current code only has the server case.  i assume you added these additional startup methods to make testing easier.  Previously, I just started the server and then used curl to POST a replay to it.  Up to you if you want to keep them around.

howardchung commented on src/main/java/yasp/Main.java:
```diff
@@ -2,35 +2,61 @@
+	try { 
+                       
+	    if (args.length > 0 && args[0].equals("--file")) { System.exit(parse(args[1])); }
+            else if (args.length >0 && args[0].equals("--")) { System.exit(parseStream(System.in, System.out)); }
```
fix >0 spacing

howardchung commented on src/main/java/yasp/Main.java:
```diff
@@ -2,35 +2,61 @@
+            else { startServer(args); }
+	}
+	catch (Exception e) {
+	    e.printStackTrace();
```
do you need to catch the exception here?  If we have one, we should crash since the startup is incorrect.

howardchung commented on src/main/java/yasp/Main.java:
```diff
@@ -2,35 +2,61 @@
+        return 0;
     }
     
+    static int parse(String replay_file) throws Exception {
```
call this parseFile

howardchung commented on src/main/java/yasp/Main.java:
```diff
@@ -2,35 +2,61 @@
+        return 0;
     }
     
+    static int parse(String replay_file) throws Exception {
```
use camelCase

howardchung replied:
and call it fileName

howardchung commented on src/main/java/yasp/Main.java:
```diff
@@ -2,35 +2,61 @@
-        server.createContext("/", new MyHandler());
-        server.setExecutor(java.util.concurrent.Executors.newFixedThreadPool(4));
-        server.start();
+    static int parseStream(InputStream is, OutputStream os) throws IOException {
```
I think methods should have visibility declared.

howardchung commented on src/main/java/yasp/Main.java:
```diff
@@ -2,35 +2,61 @@
+	catch (Exception e)
+	{
+	    e.printStackTrace();
+            return -1;
```
extra space

howardchung commented on src/main/java/yasp/Parse.java:
```diff
@@ -291,15 +295,11 @@ public void onCombatLogEntry(Context ctx, CombatLogEntry cle) {
     @OnEntityLeft
-    public void onEntityLeft(Context ctx, Entity e) {
-        processEntity(ctx, e, true);
+    public void onEntityExistenceChanged(Context ctx, Entity e) {
```
Since the annotation here is still OnEntityLeft, I think the method name should reflect that.

howardchung replied:
Actually, if this is only here for cosmetics entities, you should probably just keep the OnEntityEntered and remove OnEntityLeft (assuming ward-related annotated methods will be called separately)

micaelbergeron replied:
Yeah that could be a good way to catch this.

howardchung commented on src/main/java/yasp/Parse.java:
```diff
@@ -459,6 +459,19 @@ public void onTickStart(Context ctx, boolean synthetic) {
     }
+    
+    @OnWardCountered
+    public void onWardCountered(Context ctx, Entity e, String killer_hero_name) {
```
camelCase

howardchung commented on src/main/java/yasp/Parse.java:
```diff
@@ -459,6 +459,19 @@ public void onTickStart(Context ctx, boolean synthetic) {
+    
+    @OnWardExpired
+    @OnWardPlaced
+    public void onWardExistenceChanged(Context ctx, Entity e) {
```
can you unify this method with the previous one and detect from context which case it is?

micaelbergeron replied:
I don't think I can do that as the signatures are different.

howardchung commented on src/main/java/yasp/Parse.java:
```diff
@@ -476,51 +489,44 @@ public void onTickStart(Context ctx, boolean synthetic) {
+	}
+    }
+
+    Entry buildWardEntry(Context ctx, Entity e) {
```
add method visibility

howardchung commented on src/main/java/yasp/Parse.java:
```diff
@@ -476,51 +489,44 @@ public void onTickStart(Context ctx, boolean synthetic) {
+	entry.key = Arrays.toString(pos);
+	entry.ehandle = e.getHandle();
+
+	if (entry.entityleft)
```
use braces for all if statements

howardchung commented on src/main/java/yasp/processors/warding/Wards.java:
```diff
@@ -0,0 +1,188 @@
+@Provides({ OnWardCountered.class, OnWardExpired.class, OnWardPlaced.class })
+public class Wards {
+    
+    private static final Set<String> WARDS_DT_CLASSES = new HashSet<String>(Arrays.asList(
```
possibly add one for Source 1 entity names as well.

howardchung replied:
I think there's a shorter way to populate a set with elements that doesn't require Arrays.asList of a new String[].

micaelbergeron replied:
Yes but in this case .contains is O(1).

howardchung replied:
yes, you'd keep using a hashset.  Unfortunately it doesn't look like set literals are in Java 8: http://stackoverflow.com/questions/10705705/how-do-i-use-collection-literals-in-java-7

micaelbergeron replied:
I tried to find the literal notation not to avail.

howardchung commented on src/main/java/yasp/processors/warding/Wards.java:
```diff
@@ -0,0 +1,188 @@
+        }
+    } 
+        
+    @Initializer(OnWardCountered.class)
```
consider changing name to OnWardKilled since that's a more precise description of what happened (a combat log event indicating a killed ward)

micaelbergeron replied:
Seems legit.

howardchung committed:
Update postbuild.sh

Commit:
reword annotions and fix code style

Commit:
reword annotions and fix code style

howardchung commented on src/main/java/yasp/processors/warding/Wards.java:
```diff
@@ -0,0 +1,195 @@
+    
+    private static final Set<String> WARDS_TARGET_NAMES = new HashSet<String>(Arrays.asList(
+        new String[] {
+            "npc_dota_observer_wards",
```
are you sure it's supposed to be plural?

micaelbergeron replied:
`{"time":22,"type":"DOTA_COMBATLOG_DEATH","value":0,"attackername":"npc_dota_hero_abaddon","targetname":"npc_dota_observer_wards","sourcename":"npc_dota_hero_abaddon","targetsourcename":"npc_dota_observer_wards","attackerhero":true,"targethero":false,"attackerillusion":false,"targetillusion":false,"inflictor":"dota_unknown","gold_reason":0,"xp_reason":0}`

Don't get me started on Valve's conventions.

howardchung committed:
Update .travis.yml

howardchung committed:
extract player_slot for cosmetics

howardchung committed:
Merge branch 'master' of github.com:yasp-dota/parser

howardchung committed:
add some owner hero entity data

howardchung committed:
update rebuild script

Commit:
Merge branch 'master' of github.com:micaelbergeron/parser

Commit:
correct an edge case on same-tick events

upon a same-tick event, there is a chance that the ward kill will be credited to the wrong attacker, but this commit limits the impact of this caveat by separating the wards from the sentries.

thus you will always be credited a ward kill if you last hit a ward

Commit:
update the Dockerfile to expose the parser's port

howardchung commented on Dockerfile:
```diff
@@ -12,4 +12,6 @@ WORKDIR /usr/src/parser
 ADD . /usr/src/parser
 RUN mvn -q -f /usr/src/parser/pom.xml clean install -U
 
+EXPOSE 5600
```
how did this work before?  port option passed via Docker command line?

micaelbergeron replied:
This is because the new docker-compose is using named services to access the other services instead of going through the host networking. It is good practice to EXPOSE listening ports on your docker components, it makes it easy to expose the service to other components.

As you can see in `docker-compose.yml`

```
    environment:
      PARSER_HOST: http://odota-parser:5600
      POSTGRES_URL: postgresql://postgres:postgres@odota-postgres/yasp
      POSTGRES_TEST_URL: postgresql://postgres:postgres@odota-postgres/yasp_test
      READONLY_POSTGRES_URL: postgresql://readonly:readonly@odota-postgres/yasp
      REDIS_URL: redis://odota-redis:6379/0
      REDIS_TEST_URL: redis://odota-redis:6379/1
      CASSANDRA_URL: cassandra://odota-cassandra/yasp
      CASSANDRA_TEST_URL: cassandra://odota-cassandra/yasp_test
```

I think docker-compose automatically expose the image EXPOSED ports.

howardchung commented on src/main/java/yasp/processors/warding/Wards.java:
```diff
@@ -0,0 +1,210 @@
+        WARDS_TARGET_NAMES = Collections.unmodifiableSet(new HashSet<>(target_by_dtclass.values()));
+    }
+    
+    // simple inderection map
```
indirection?

micaelbergeron replied:
Opps, that's not relevant anymore.

howardchung commented on Dockerfile:
```diff
@@ -12,4 +12,6 @@ WORKDIR /usr/src/parser
 
+EXPOSE 5600
+
 CMD ["java", "-jar", "-Xmx256m", "/usr/src/parser/target/stats-0.1.0.jar", "5600"]
```
will you need to update this for the jar name?

micaelbergeron replied:
No, I've changed the pom.xml to output the same filename as before.

Commit:
remove the indirection map (was not used)

dartharaknus commented on src/main/java/yasp/Parse.java:
```diff
@@ -476,51 +504,45 @@ public void onTickStart(Context ctx, boolean synthetic) {
+
+    private Entry buildWardEntry(Context ctx, Entity e) {
+	Entry entry = new Entry(time);
+        boolean isObserver = !e.getDtClass().getDtName().contains("TrueSight");
```
Extra space
